File: python/Sliding Window/best_time_buy_sell_stock.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """
        Find the maximum profit from a single buy and sell of stock.
        Approach:
        - Use two pointers (buy and sell) to track the minimum price and the best profit.
        - Move the sell pointer forward, updating buy pointer if a lower price is found.
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        l, r = 0, 1  # l = buy pointer, r = sell pointer
        max_profit = 0
        
        # Slide the window until we reach the end
        while r < len(prices):
            # If current price is higher than buy price, calculate profit
            if prices[l] < prices[r]:
                profit = prices[r] - prices[l]
                max_profit = max(profit, max_profit)
            else:
                # If current price is lower, update buy pointer
                # This is because we found a better buying point
                l = r
            r += 1  # Always move sell pointer forward
        
        return max_profit

# -------------------
# Test Cases
# -------------------
def test_max_profit():
    s = Solution()
    # Test case 1: Increasing prices
    assert s.maxProfit([1,2,3,4,5]) == 4
    # Test case 2: Decreasing prices
    assert s.maxProfit([5,4,3,2,1]) == 0
    # Test case 3: Fluctuating prices
    assert s.maxProfit([7,1,5,3,6,4]) == 5
    # Test case 4: Single price
    assert s.maxProfit([5]) == 0
    # Test case 5: Empty list
    assert s.maxProfit([]) == 0
    # Test case 6: Two prices, profit
    assert s.maxProfit([1,2]) == 1
    # Test case 7: Two prices, no profit
    assert s.maxProfit([2,1]) == 0
    print("All test cases passed!")

if __name__ == "__main__":
    test_max_profit()

File: C++/Sliding Window/best_time_buy_sell_stock.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Problem: Find maximum profit from single buy-sell stock transaction
     * 
     * Approach:
     * - Use sliding window with two pointers (buy and sell)
     * - Left pointer tracks minimum price seen
     * - Right pointer looks for selling opportunities
     * - Update max profit when higher selling price found
     * - Move buy pointer when lower price found
     * 
     * Time: O(n) - single pass through array
     * Space: O(1) - only using two pointers and variables
     */
    int maxProfit(vector<int>& prices) {
        // Edge case: empty array or single price
        if (prices.size() < 2) return 0;
        
        // Initialize pointers and max profit
        int buy = 0;          // Buy pointer (left)
        int sell = 1;         // Sell pointer (right)
        int maxProfit = 0;    // Track maximum profit
        
        // Continue until sell pointer reaches end
        while (sell < prices.size()) {
            // If profitable, calculate profit
            if (prices[buy] < prices[sell]) {
                int profit = prices[sell] - prices[buy];
                maxProfit = max(maxProfit, profit);
            }
            else {
                // Found lower price, update buy pointer
                buy = sell;
            }
            // Always move sell pointer forward
            sell++;
        }
        
        return maxProfit;
    }
};

/*
 * Detailed C++ Syntax Explanation for Beginners:
 * 
 * 1. Vector Operations:
 *    prices.size()     // Get array length
 *    prices[index]     // Access element at index
 * 
 * 2. Variable Declarations:
 *    int buy = 0      // Buy pointer at start
 *    int sell = 1     // Sell pointer one step ahead
 *    int maxProfit = 0  // Track best profit found
 * 
 * 3. Standard Library:
 *    max(a, b)        // Returns larger of two values
 * 
 * 4. Edge Case Check:
 *    if (prices.size() < 2)  // Check if array too small
 *    return 0                // Return early if invalid
 * 
 * 5. Profit Calculation:
 *    prices[sell] - prices[buy]  // Current profit
 *    max(maxProfit, profit)      // Update if better
 * 
 * 6. While Loop:
 *    while (sell < prices.size())  // Continue until end
 *    sell++                        // Move window forward
 * 
 * Note: This implementation:
 * - Uses O(n) time by processing each price once
 * - Uses O(1) space with only variables
 * - Handles edge cases properly
 */
